### PR TITLE
Add scaled_mm_v2 cpu implementation

### PR DESCRIPTION
**Pull request by CWOA**
This PR adds an emulated CPU implementation of `scaled_mm_v2` for tensor- and row-wise scaling and registers it. The existing V1 function also routes into this emulated implementation.

This PR is the third part of a series:
- https://github.com/pytorch/pytorch/pull/175657
- https://github.com/pytorch/pytorch/pull/175658

which was first put forward in: https://github.com/pytorch/pytorch/pull/170670

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @voznesenskym @penguinwu @EikanWang @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: torch/testing/_internal/common_cuda.py [hunk 1]
```diff
@@ -144,7 +144,8 @@ def evaluate_platform_supports_fp8():
                     return True
         else:
             return SM90OrLater or torch.cuda.get_device_capability() == (8, 9)
-    return False
+    # As CPU supports FP8 and is always available, return True.
```
Comment (slayton58) [inline torch/testing/_internal/common_cuda.py:176]: Is this actually true? Are fp8 paths available on all CPUs (emulated or otherwise)?

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -8968,7 +8968,7 @@ def sample_inputs_scaled_mm_v2(op_info, device, dtype, requires_grad, **kwargs):
     mat1 = make_mat_e4m3((M, K))
     mat2 = make_mat_e4m3((K, N)).t().contiguous().t()
 
-    if torch.device(device).type == "cuda":
+    if torch.cuda.is_available() and torch.device(device).type == "cuda":
```
Comment (slayton58) [inline torch/testing/_internal/common_methods_invocations.py]: Necessary?

File: aten/src/ATen/native/ScaledBlas.cpp [hunk 3]
```diff
@@ -140,6 +322,197 @@ _scaled_mm_cpu(const Tensor& mat_a, const Tensor& mat_b,
...
+
+  // Conversion of implicitly-defined enums to explicit
+  auto scale_recipe_a_enum = convert_int_to_enum<ScalingType>(scale_recipe_a);
+  auto swizzle_a_enum = convert_int_to_enum<SwizzleType>(swizzle_a);
+  auto scale_recipe_b_enum = convert_int_to_enum<ScalingType>(scale_recipe_b);
+  auto swizzle_b_enum = convert_int_to_enum<SwizzleType>(swizzle_b);
+
+  // CPU does not support swizzle, so return false.
+  TORCH_CHECK_VALUE(
+      swizzle_a_enum[0] == at::blas::SwizzleType::NO_SWIZZLE &&
```
Comment (slayton58) [inline aten/src/ATen/native/ScaledBlas.cpp:424]: I don't see bounds checking before this line - what if the user passes `swizzle_a=[]`?

File: aten/src/ATen/native/ScaledBlas.cpp [hunk 4]
```diff
@@ -96,7 +255,7 @@ _scaled_mm_out_cpu_emulated(const Tensor& mat1, const Tensor& mat2,
 }
 
 Tensor&
-_scaled_mm_out_cpu(const Tensor& mat1, const Tensor& mat2,
+_scaled_mm_out_cpu(const Tensor& mat_a, const Tensor& mat_b,
```
Comment (slayton58) [inline aten/src/ATen/native/ScaledBlas.cpp]: Necessary to change these? Seems like it just clutters the PR a little.

File: aten/src/ATen/native/ScaledBlas.cpp [hunk 5]
```diff
@@ -32,57 +33,215 @@
...
           const Tensor& scale_b,
           const std::optional<at::Tensor>& bias,
           const std::optional<at::Tensor>& scale_result,
           std::optional<c10::ScalarType> out_dtype,
           bool use_fast_accum,
+          const scaled::ScaledGemmImplementation gemm_impl,
           Tensor& out) {
-  TORCH_CHECK(mat1.dim() == 2, "mat1 must be a matrix");
-  TORCH_CHECK(mat2.dim() == 2, "mat2 must be a matrix");
+  TORCH_CHECK(mat_a.dim() == 2, "mat_a must be a matrix");
```
Comment (slayton58) [inline aten/src/ATen/native/ScaledBlas.cpp]: Same comment w.r.t. `mat{1,2} -> mat_{a,b}` being a necessary change

File: aten/src/ATen/native/ScaledBlas.cpp [hunk 6]
```diff
@@ -140,6 +322,197 @@ _scaled_mm_cpu(const Tensor& mat_a, const Tensor& mat_b,
...
+
+    if (ok) {
+      gemm_impl = scaled_gemm_impl;
+      found_impl = true;
+      break;
+    }
+  }
+
+  if (!found_impl) {
+    invalid_scaling_config(mat_a, mat_b, scale_a[0], scale_b[0]);
```
Comment (slayton58) [inline aten/src/ATen/native/ScaledBlas.cpp]: Bounds checking? If user passes `scale_a=[]` what happens? It'll fail all acceptance functions, but if it manages to get here, it'll OOB access..

File: aten/src/ATen/native/ScaledBlas.cpp [hunk 7]
```diff
@@ -140,6 +322,197 @@ _scaled_mm_cpu(const Tensor& mat_a, const Tensor& mat_b,
...
+
+  if (gemm_impl == ScaledGemmImplementation::TENSORWISE_TENSORWISE ||
+      gemm_impl == ScaledGemmImplementation::ROWWISE_ROWWISE) {
+    _scaled_mm_out_cpu_emulated(
+      mat_a,
+      mat_b,
+      scale_a[0],
+      scale_b[0],
+      bias,
+      std::nullopt,
```
Comment (slayton58) [inline aten/src/ATen/native/ScaledBlas.cpp]: Nit: Note what the `std::nullopt` is referring to
```suggestion
      std::nullopt,  // scale-result
```

**Review by slayton58 — COMMENTED** [review]
Small nits, but LGTM :)

File: aten/src/ATen/native/ScaledBlas.cpp [hunk 8]
```diff
@@ -32,12 +33,102 @@
...
 #endif
 
 namespace at::native {
 
+using at::blas::ScalingType;
+using at::blas::SwizzleType;
+
+namespace {
+
+void invalid_scaling_config(
```
Comment (drisspg) [inline aten/src/ATen/native/ScaledBlas.cpp:48]: nit: @slayton58 is this the same pattern you have, kind feels nice if you could have the error messages more collocated to the `is_x_scaling`

File: torch/testing/_internal/common_cuda.py (hunk 1 above)
Comment (drisspg) [reply torch/testing/_internal/common_cuda.py:176]: yeah I think this is true its just a tensor dtype with the conversion functions defined and as long as the impl casts up before doing any math should be good

File: aten/src/ATen/native/ScaledBlas.cpp (hunk 8 above)
Comment (slayton58) [reply aten/src/ATen/native/ScaledBlas.cpp:48]: This is the general pattern used, yeah - I think there's a solid argument for refactoring all the error messages to something more pleasant & easy to maintain.

Some form of collocating would be nice, maybe `not_x_scaling_reason()` or similar, so we can iterate at the end over the available options (nothing that `is_x_scaling(..)` returning `false` is a non-terminating condition, we're just evaluating if that particular scaling is desired, not that there are no valid scalings available, which is when this kind of message is printed)

File: test/test_mkldnn.py [hunk 9]
```diff
@@ -1647,7 +1647,7 @@ def test_mkldnn_error_on_zero_stride(self, device):
 
     def test_mkldnn_scaled_mm(self, device) -> None:
         # test with input scale, weight scale and output_scale
-        M, N, K = 2, 13, 16
+        M, N, K = 2, 16, 16
```
Comment (gujinghui) [inline test/test_mkldnn.py]: Why we have to change the shape? This case is only for CPU, or will also impact XPU?

Comment (CWOA) [reply test/test_mkldnn.py]: Because otherwise the size check [here](https://github.com/graphcore/pytorch-fork/blob/b8c5d16bff548fe182a5df60079b0f778c0e8f8b/torch/_meta_registrations.py#L6645) fails as I've added cpu to the [size checks](https://github.com/graphcore/pytorch-fork/blob/b8c5d16bff548fe182a5df60079b0f778c0e8f8b/torch/_meta_registrations.py#L6621)

Comment (slayton58) [reply test/test_mkldnn.py]: Is this an actual requirement of all the CPU code paths? It would seem to indicate a regression in functionality - CPU could previously handle `shape % 16 != 0` and now it can't

Comment (CWOA) [reply test/test_mkldnn.py]: Have restricted the checks to non-cpu devices: 713e9e970601a7aced1e5a228bfbaf37de620fff

Comment (slayton58) [reply test/test_mkldnn.py]: But what _should_ be checked? What are the restrictions on the CPU side after this PR?

**Review by slayton58 — APPROVED** [review]
Thanks for bearing with the reviews - LGTM!

File: torch/testing/_internal/common_cuda.py (hunk 1 above)
Comment (jeffdaily) [reply torch/testing/_internal/common_cuda.py:176]: @jagadish-amd pointed out to me that the HIP logic falls through to this `False`, and this change makes it always `True` for HIP.  He'll have a forward fix coming soon.  It broke our MI200 tests, but other CI issues masked the problem.
